modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java

```diff
@@ -1951,6 +1959,86 @@ private void linkFunction(FunctionInvocationExpr funcIExpr) {
         funcIExpr.setCallableUnit(function);
     }
 
+    /**
+     * Helper method to find the best function match when there is no direct match.
+     *
+     * @param symbolName
+     * @return bLangSymbol
+     */
+    private BLangSymbol findBestMatchForFunctionSymbol(FunctionInvocationExpr funcIExpr, SymbolName symbolName) {
+        BLangSymbol functionSymbol = null;
+        BLangSymbol pkgSymbol = null;
+        if (symbolName.getPkgPath() == null) {
+            pkgSymbol = (BLangPackage) getCurrentPackageScope(currentScope);
+        } else {
+            SymbolName pkgSymbolName = new SymbolName(symbolName.getPkgPath());
+            pkgSymbol = currentScope.resolve(pkgSymbolName);
+        }
+        //pkgSymbol should be a instance of SymbolScope, hence doesn't do the check here
+        if (pkgSymbol == null) {
+            return null;
+        }
+        for (Map.Entry entry : ((SymbolScope) pkgSymbol).getSymbolMap().entrySet()) {
+            if (!(entry.getKey() instanceof FunctionSymbolName)) {
+                continue;
+            }
+            FunctionSymbolName funcSymName = (FunctionSymbolName) entry.getKey();
+            if (!funcSymName.isFunctionsEqual((FunctionSymbolName) symbolName)) {
+                continue;
+            }
+
+            boolean implicitCastPossible = true;
+            Expression[] exprs = funcIExpr.getArgExprs();
+            for (int i = 0; i < exprs.length; i++) {
+                BType lhsType = ((Function) entry.getValue()).getParameterDefs()[i].getType();
+
+                BType rhsType = exprs[i].getType();
+                if (rhsType != null && lhsType.equals(rhsType)) {
+                    continue;
+                }
+                TypeCastExpression newExpr = checkWideningPossible(lhsType, exprs[i]);
+                if (newExpr != null) {
+                    exprs[i] = newExpr;
+                } else {
+                    implicitCastPossible = false;
+                    break;
+                }
+            }
+            if (implicitCastPossible && functionSymbol == null) {
+                functionSymbol = (BLangSymbol) entry.getValue();
+            } else {
```

Comment:
There is a bug in this code.  Assume I want to invoke the function "print(float)", but only the following functions are defined in the symbol map.

print(string)
print(int)
print(Person)
print(Employee)

This code will try to throw an "ambiguous function error ".